### PR TITLE
Add obstacle detection sensor

### LINKED ISSUE #1343: Extended sensor suites (obstacle, dynamic object, red-light etc)
Implement a suite of sensors for flexible combination for different purposed.
For example, they can be used to construct a structured observation for learning.

1. Red-light sensor: https://github.com/HorizonRobotics/alf/pull/1220
2. Obstacle sensor https://github.com/HorizonRobotics/alf/pull/1344

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
Implemented obstacle detection sensor for CARLA.
This sensor is used in the imitation learning experiment.

### CODE REVIEW

File: alf/environments/carla_sensors.py [hunk 1]
```diff
@@ -2050,3 +2050,74 @@ def _world_to_pixel(self, location: carla.Location, projective=False):
...
+
+
+# ==============================================================================
+# -- Obstacle Detection Sensor -------------------------------------------------
+# ==============================================================================
+
+
+@alf.configurable
+class ObstacleDetectionSensor(SensorBase):
+    """ObstacleDetectionSensor.
```
Comment (breakds) [inline alf/environments/carla_sensors.py:2062]: I think it is still good to document arguments about `hit_radius` and `only_dynamics` as the name does not speak much about them.

File: alf/environments/carla_sensors.py [hunk 2]
```diff
@@ -2050,3 +2050,74 @@ def _world_to_pixel(self, location: carla.Location, projective=False):
...
+        bp.set_attribute('hit_radius', str(hit_radius))
+        bp.set_attribute('debug_linetrace', 'false')
+        if only_dynamics:
+            bp.set_attribute('only_dynamics', 'true')
+        else:
+            bp.set_attribute('only_dynamics', 'false')
+        # Put this sensor on the windshield of the actor to avoid detection
+        # of collision with the actor itself
+        self._sensor_transform = carla.Transform(
+            carla.Location(x=2.0, z=1.7), carla.Rotation(yaw=0))
```
Comment (breakds) [inline alf/environments/carla_sensors.py]: Is this windshield offset universal to all cars or only to the car model we are using? If so, accidental change of car model might cause a surprise.

File: alf/environments/carla_sensors.py [hunk 3]
```diff
@@ -2050,3 +2050,74 @@ def _world_to_pixel(self, location: carla.Location, projective=False):
...
+
+    @staticmethod
+    def _on_obstacle(weak_self, event):
+        self = weak_self()
+        if not self:
+            return
+        self._distance = event.distance
+        self._event_count += 1
+        # the message is useful for debugging purpose
+        self._msg = ("Event %s, in line of sight with %s at distance %u" % (
```
Comment (breakds) [inline alf/environments/carla_sensors.py]: Curious where the message will go? Log?

File: alf/environments/carla_sensors.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/environments/carla_sensors.py:2062]: Good point. Now added more info on these arguments.

File: alf/environments/carla_sensors.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/environments/carla_sensors.py]: Changed this to an argument in order to accommodate very different sized vehicles and added document for its value setting.

File: alf/environments/carla_sensors.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/environments/carla_sensors.py]: Now added optional logging for the message.

**Review by breakds — APPROVED** [review]

**Pull request merged**